src/main/java/com/gw2auth/oauth2/server/service/application/client/account/ApplicationClientAccount.java

```diff
@@ -19,7 +23,7 @@ public static ApplicationClientAccount fromEntity(ApplicationClientAccountEntity
                 entity.applicationId(),
                 ApprovalStatus.valueOf(entity.approvalStatus()),
                 entity.approvalRequestMessage(),
-                entity.authorizedScopes()
+                entity.authorizedScopes().stream().map(OAuth2Scopes::fromOAuth2).flatMap(Optional::stream).collect(Collectors.toUnmodifiableSet())
```

Comment:
If you want to have defensive copies here, it is better to put them in a constructor instead:

```java 
public ApplicationClientAccount {
    this.authorizedScopes = authorizedScopes.stream().map(OAuth2Scopes::fromOAuth2).flatMap(Optional::stream).collect(Collectors.toUnmodifiableSet())
}
```

Reply:
Generally I agree this would be a good approach, however the generic type changes here:

- `Set<String>` (persistence level)
- `Set<OAauth2Scope>` (business level)

all these Sets floating around between the different kind of DTOs are deeply immutable (unmodifiable sets everywhere - sadly no guarantee for this by type, but adding another defensive copy (of the set) to the constructor if all creations go through the factory anyway feels a bit too much)